Fingerprint fail on generating TMD

It looks like the [tmd_file_fingerprint.json](https://github.com/PSLmodels/tax-microdata-benchmarking/blob/master/tests/fingerprints/tmd_file_fingerprint.json#L881) needs to be updated. Errors are being thrown when generating the TMD with TC >=6.6.1. I don't fully understand the other FAILED or else I would make the change to tmd_file_fingerprint.json:

Output from recent generation.:

FAILED tests/test_tax_expenditures.py::test_tax_exp_diffs - ValueError: gfact...
FAILED tests/test_tmd_file_fingerprint.py::test_tmd_file_fingerprint - Failed...
FAILED tests/test_weights.py::test_weights - ValueError: WEIGHT FINGERPRINT M...
FAILED tests/test_imputed_variables.py::test_obbba_deduction_tax_benefits - V...

@martinholmer  - thank you for the help on this. Below is the output from the latest regeneration. While looking at this, I tracked down that some issues were caused by not gzipping the CEX23 and SIPP24 input files, which was causing some imputations to be zero (makes sense!). That was solely a me problem. 

But I am still getting the four failures from before. Using Tax-Calc 6.6.2 on Windows 11, Python 3.12.13. This same error is occurring on my other Windows machine and a co-authors, so I don't think its a "me" issue, but please let me know if this _does_ look like user error. I think the main issues is the one in bold and has to do with Windows vs Mac, but I am getting WAY out of my comfort zone so I called in backup and its suggestion is below as well. 

Please let me me know if I can provide more (or less) info. 

Thank you!
Mark 
 
The exact differences are:

**FAILED tests/test_tax_expenditures.py::test_tax_exp_diffs - ValueError: gfactors is not a GrowFactors instance**
FAILED tests/test_tmd_file_fingerprint.py::test_tmd_file_fingerprint - Failed: 34 statistic mismatch(es) in tmd.csv.gz fingerprint:
FAILED tests/test_weights.py::test_weights - ValueError: WEIGHT FINGERPRINT MISMATCH USING 2022 DATA:
FAILED tests/test_imputed_variables.py::test_obbba_deduction_tax_benefits - ValueError:

**Issue:**

Root cause identified for `test_tax_expenditures` failure.

The issue is in `tmd/utils/taxcalc_output.py`. The path type checks use `pathlib.PosixPath` which only matches on Mac/Linux. On Windows the path is a `WindowsPath`, so it falls through to the `else` branch and passes the raw path object directly to `taxcalc.Records` instead of converting it to a `GrowFactors` instance first.

Source of `add_taxcalc_outputs` as it runs on Windows:

```python
if isinstance(growfactors, pathlib.PosixPath):
    growf = taxcalc.GrowFactors(growfactors_filename=str(growfactors))
else:
    growf = growfactors  # on Windows, growfactors is a WindowsPath, not a GrowFactors instance
```

Confirmed with:

```bash
python -c "import taxcalc; from pathlib import Path; gf = Path('tmd/storage/output/tmd_growfactors.csv'); print(type(gf)); rec = taxcalc.GrowFactors(str(gf)); print(rec)"
```

Output:

```
<class 'pathlib.WindowsPath'>
<taxcalc.growfactors.GrowFactors object at 0x000001E1BD3D5BE0>
```

`taxcalc.GrowFactors` works correctly when the path is passed as a string. The problem is that `taxcalc_output.py` checks for `PosixPath` only, so the conversion never happens on Windows.

**Suggested fix** — change both `PosixPath` checks to `pathlib.Path` in `taxcalc_output.py`:

```python
if isinstance(growfactors, pathlib.Path):  # was PosixPath
    growf = taxcalc.GrowFactors(growfactors_filename=str(growfactors))

if isinstance(weights, pathlib.Path):  # was PosixPath
    wghts = str(weights)
```

`pathlib.Path` is the base class for both `PosixPath` and `WindowsPath` and is the correct cross-platform check.




$ make data
pip install -e .
Obtaining file:///C:/Users/<user>/Modeling/PSL/tax-microdata-benchmarking
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: taxcalc>=6.6.1 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (6.6.2)
Requirement already satisfied: numpy in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (2.4.4)
Requirement already satisfied: pandas>=3.0.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (3.0.2)
Requirement already satisfied: clarabel in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (0.11.1)
Requirement already satisfied: scikit-learn in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (1.9.0)
Requirement already satisfied: scipy in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (1.17.1)
Requirement already satisfied: xlrd in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (2.0.2)
Requirement already satisfied: openpyxl in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (3.1.5)
Requirement already satisfied: tqdm in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (4.67.3)
Requirement already satisfied: tables in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (3.11.1)
Requirement already satisfied: requests in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (2.33.1)
Requirement already satisfied: PyYAML in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (6.0.3)
Requirement already satisfied: black>=26.1.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (26.5.1)
Requirement already satisfied: pycodestyle>=2.14.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (2.14.0)
Requirement already satisfied: pylint>=3.3.8 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (4.0.5)
Requirement already satisfied: pytest in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (9.0.3)
Requirement already satisfied: pytest-xdist in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tmd==2.1.1) (3.8.0)
Requirement already satisfied: click>=8.0.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from black>=26.1.0->tmd==2.1.1) (8.4.1)
Requirement already satisfied: mypy-extensions>=0.4.3 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from black>=26.1.0->tmd==2.1.1) (1.1.0)
Requirement already satisfied: packaging>=22.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from black>=26.1.0->tmd==2.1.1) (26.0)
Requirement already satisfied: pathspec>=1.0.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from black>=26.1.0->tmd==2.1.1) (1.1.1)
Requirement already satisfied: platformdirs>=2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from black>=26.1.0->tmd==2.1.1) (4.10.0)
Requirement already satisfied: pytokens~=0.4.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from black>=26.1.0->tmd==2.1.1) (0.4.1)
Requirement already satisfied: colorama in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from click>=8.0.0->black>=26.1.0->tmd==2.1.1) (0.4.6)
Requirement already satisfied: python-dateutil>=2.8.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pandas>=3.0.2->tmd==2.1.1) (2.9.0.post0)
Requirement already satisfied: tzdata in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pandas>=3.0.2->tmd==2.1.1) (2026.1)
Requirement already satisfied: astroid<=4.1.dev0,>=4.0.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pylint>=3.3.8->tmd==2.1.1) (4.0.4)
Requirement already satisfied: dill>=0.3.6 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pylint>=3.3.8->tmd==2.1.1) (0.4.1)
Requirement already satisfied: isort!=5.13,<9,>=5 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pylint>=3.3.8->tmd==2.1.1) (8.0.1)
Requirement already satisfied: mccabe<0.8,>=0.6 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pylint>=3.3.8->tmd==2.1.1) (0.7.0)
Requirement already satisfied: tomlkit>=0.10.1 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pylint>=3.3.8->tmd==2.1.1) (0.15.0)
Requirement already satisfied: six>=1.5 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from python-dateutil>=2.8.2->pandas>=3.0.2->tmd==2.1.1) (1.17.0)
Requirement already satisfied: bokeh>=2.4 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from taxcalc>=6.6.1->tmd==2.1.1) (3.9.0)
Requirement already satisfied: numba in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from taxcalc>=6.6.1->tmd==2.1.1) (0.65.0)
Requirement already satisfied: paramtools>=0.20.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from taxcalc>=6.6.1->tmd==2.1.1) (0.20.0)
Requirement already satisfied: behresp in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from taxcalc>=6.6.1->tmd==2.1.1) (0.11.0)
Requirement already satisfied: Jinja2>=2.9 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (3.1.6)
Requirement already satisfied: contourpy>=1.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (1.3.3)
Requirement already satisfied: narwhals>=1.13 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (2.20.0)
Requirement already satisfied: pillow>=7.1.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (12.2.0)
Requirement already satisfied: tornado>=6.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (6.5.5)
Requirement already satisfied: xyzservices>=2021.09.1 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (2026.3.0)
Requirement already satisfied: MarkupSafe>=2.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from Jinja2>=2.9->bokeh>=2.4->taxcalc>=6.6.1->tmd==2.1.1) (3.0.3)
Requirement already satisfied: marshmallow>=4.0.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from paramtools>=0.20.0->taxcalc>=6.6.1->tmd==2.1.1) (4.3.0)
Requirement already satisfied: fsspec in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from paramtools>=0.20.0->taxcalc>=6.6.1->tmd==2.1.1) (2026.3.0)
Requirement already satisfied: sortedcontainers in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from paramtools>=0.20.0->taxcalc>=6.6.1->tmd==2.1.1) (2.4.0)
Requirement already satisfied: cffi in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from clarabel->tmd==2.1.1) (2.0.0)
Requirement already satisfied: pycparser in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from cffi->clarabel->tmd==2.1.1) (3.0)
Requirement already satisfied: llvmlite<0.48,>=0.47.0dev0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from numba->taxcalc>=6.6.1->tmd==2.1.1) (0.47.0)
Requirement already satisfied: et-xmlfile in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from openpyxl->tmd==2.1.1) (2.0.0)
Requirement already satisfied: iniconfig>=1.0.1 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pytest->tmd==2.1.1) (2.3.0)
Requirement already satisfied: pluggy<2,>=1.5 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pytest->tmd==2.1.1) (1.6.0)
Requirement already satisfied: pygments>=2.7.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pytest->tmd==2.1.1) (2.20.0)
Requirement already satisfied: execnet>=2.1 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pytest-xdist->tmd==2.1.1) (2.1.2)
Requirement already satisfied: charset_normalizer<4,>=2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from requests->tmd==2.1.1) (3.4.7)
Requirement already satisfied: idna<4,>=2.5 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from requests->tmd==2.1.1) (3.11)
Requirement already satisfied: urllib3<3,>=1.26 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from requests->tmd==2.1.1) (2.6.3)
Requirement already satisfied: certifi>=2023.5.7 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from requests->tmd==2.1.1) (2026.2.25)
Requirement already satisfied: joblib>=1.4.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from scikit-learn->tmd==2.1.1) (1.5.3)
Requirement already satisfied: threadpoolctl>=3.5.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from scikit-learn->tmd==2.1.1) (3.6.0)
Requirement already satisfied: numexpr>=2.6.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tables->tmd==2.1.1) (2.14.1)
Requirement already satisfied: py-cpuinfo in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tables->tmd==2.1.1) (9.0.0)
Requirement already satisfied: blosc2>=2.3.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from tables->tmd==2.1.1) (4.4.1)
Requirement already satisfied: ndindex in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from blosc2>=2.3.0->tables->tmd==2.1.1) (1.10.1)
Requirement already satisfied: msgpack in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from blosc2>=2.3.0->tables->tmd==2.1.1) (1.1.2)
Requirement already satisfied: pydantic in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from blosc2>=2.3.0->tables->tmd==2.1.1) (2.13.4)
Requirement already satisfied: annotated-types>=0.6.0 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pydantic->blosc2>=2.3.0->tables->tmd==2.1.1) (0.7.0)
Requirement already satisfied: pydantic-core==2.46.4 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pydantic->blosc2>=2.3.0->tables->tmd==2.1.1) (2.46.4)
Requirement already satisfied: typing-extensions>=4.14.1 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pydantic->blosc2>=2.3.0->tables->tmd==2.1.1) (4.15.0)
Requirement already satisfied: typing-inspection>=0.4.2 in C:\Users\<user>\anaconda3\envs\taxcalc-env\Lib\site-packages (from pydantic->blosc2>=2.3.0->tables->tmd==2.1.1) (0.4.2)
Building wheels for collected packages: tmd
  Building editable for tmd (pyproject.toml) ... done
  Created wheel for tmd: filename=tmd-2.1.1-0.editable-py3-none-any.whl size=2769 sha256=af10cc37dabcb183b4f013c19742598fa9f0ce944eac12d327bfd1bf5a70fb4c
  Stored in directory: C:\Users\<user>\AppData\Local\Temp\pip-ephem-wheel-cache-sktdz9d5\wheels\f0\9f\f6\32a99211068b1a72fef85226205325ad5f4426400d5aa85ae3
Successfully built tmd
Installing collected packages: tmd
  Attempting uninstall: tmd
    Found existing installation: tmd 2.1.1
    Uninstalling tmd-2.1.1:
      Successfully uninstalled tmd-2.1.1
Successfully installed tmd-2.1.1
rm -f tmd/storage/output/tmd*
rm -f tmd/storage/output/cached*
rm -f tmd/storage/output/preimpute_tmd.csv.gz
rm -f tmd/storage/output/make_data_*.log
black . -l 79
All done! ✨ 🍰 ✨
63 files left unchanged.
python -u tmd/create_taxcalc_input_variables.py 2>&1 \
            | tee tmd/storage/output/make_data_tmd.log
Creating 2022 PUF+CPS file assuming:
  IMPUTATION_RF_RNG_SEED = 1928374
  IMPUTATION_BETA_RNG_SEED = 37465
  ASSUMED SALT_GROW_RATE = 0.047
  ASSUMED ITMDED_GROW_RATE = 0.020
  ASSUMED W2_WAGES_RATIO = 0.15500
  ASSUMED CPS_WEIGHTS_SCALE = 1.00
Reading raw 2015 PUF data...
Projecting PUF data from 2015 to 2022...
Pre-processing PUF data...
Imputing missing PUF demographics...
Computing earnings splits...
Imputing pension contributions...
Building Tax-Calculator input dataframe for 2022...
Creating CPS dataframe for year 2022...
Combining PUF filers and CPS nonfilers...
Adding Tax-Calculator outputs for 2022...
Reweighting...
...constrained QP reweighting for year 2022
...constraint tolerance: +-0.5%
...multiplier bounds: [0.1, 10.0]
...prescale factor: 0.997790 (target=161,336,659, current=161,694,029)
...targets: 550, records: 215494
...constraint matrix: 215494x550, nnz=2,706,652, density=2.28%
...starting solver
...solve time: 13.7s (21 iterations)
...solver status: Solved
...target accuracy (550 targets):
    mean |relative error|: 0.004547
    max  |relative error|: 0.005000
    worst targets:
        0.500% | target=           1514 | achieved=           1522 | count/count/AGI in [10.0m, inf)/all returns/Married Filing Separately
        0.500% | target=           4837 | achieved=           4861 | count/count/AGI in [10.0m, inf)/all returns/Single
        0.500% | target=          30987 | achieved=          30832 | partnership and s corp income/count/AGI in [10.0m, inf)/all returns/All
        0.500% | target=           6948 | achieved=           6913 | total social security/count/AGI in [10.0m, inf)/all returns/All
        0.500% | target=           1649 | achieved=           1657 | count/count/AGI in [5.0m, 10.0m)/all returns/Married Filing Separately
        0.500% | target=          34630 | achieved=          34803 | count/count/AGI in [10.0m, inf)/all returns/All
        0.500% | target=           7631 | achieved=           7669 | total pension income/count/AGI in [10.0m, inf)/all returns/All
        0.500% | target=          25066 | achieved=          25191 | capital gains gross/count/AGI in [10.0m, inf)/all returns/All
        0.500% | target=           6098 | achieved=           6128 | count/count/AGI in [1.0m, 1.5m)/all returns/Married Filing Separately
        0.500% | target=          10500 | achieved=          10447 | total social security/count/AGI in [5.0m, 10.0m)/all returns/All
...final weights: total=189765156.11, mean=880.605289, sdev=1092.990282
Removing output variables from PUF+CPS dataframe...
Writing PUF+CPS file... [tmd/storage/output/tmd.csv.gz]
python -u tmd/create_taxcalc_sampling_weights.py 2>&1 \
            | tee tmd/storage/output/make_data_weights.log
python -u tmd/create_taxcalc_growth_factors.py 2>&1 \
            | tee tmd/storage/output/make_data_growfactors.log
python -u tmd/create_taxcalc_cached_files.py 2>&1 \
            | tee tmd/storage/output/make_data_cached.log
python -u tmd/create_taxcalc_imputed_variables.py 2>&1 \
            | tee tmd/storage/output/make_data_preimpute.log
Preparing external data for imputing missing variables...
Imputing overtime and tip data from SIPP to TMD...
Imputing auto loan interest data from CEX to TMD...
Writing preimpute TMD file... [tmd/storage/output/preimpute_tmd.csv.gz]
Writing augmented TMD file... [tmd/storage/output/tmd.csv.gz]
pytest . -v -n4 \
            --ignore=tests/national_targets_pipeline \
            --ignore=tests/test_fingerprint.py \
            --ignore=tests/test_state_weight_results.py \
            --ignore=tests/test_cd_crosswalk.py \
            --ignore=tests/test_prepare_targets.py
=========================================================================== test session starts ===========================================================================
platform win32 -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0 -- C:\Users\<user>\anaconda3\envs\taxcalc-env\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\<user>\Modeling\PSL\tax-microdata-benchmarking
configfile: pytest.ini
plugins: xdist-3.8.0
4 workers [61 items]
scheduling tests via LoadScheduling

tests/test_mice.py::TestConstructorValidation::test_min_leaf_node_size_validation
tests/test_mice.py::TestConstructorValidation::test_x_ign_validation
tests/test_mice.py::TestConstructorValidation::test_x_obs_validation
tests/test_imputed_variables.py::test_obbba_deduction_tax_benefits
[gw3] [  1%] PASSED tests/test_mice.py::TestConstructorValidation::test_min_leaf_node_size_validation
[gw2] [  3%] PASSED tests/test_mice.py::TestConstructorValidation::test_x_ign_validation
[gw1] [  4%] PASSED tests/test_mice.py::TestConstructorValidation::test_x_obs_validation
tests/test_mice.py::TestConstructorValidation::test_iters_validation
tests/test_mice.py::TestConstructorValidation::test_x_var_validation
tests/test_mice.py::TestConstructorValidation::test_seed_validation
[gw3] [  6%] PASSED tests/test_mice.py::TestConstructorValidation::test_iters_validation
[gw2] [  8%] PASSED tests/test_mice.py::TestConstructorValidation::test_seed_validation
[gw1] [  9%] PASSED tests/test_mice.py::TestConstructorValidation::test_x_var_validation
tests/test_mice.py::TestConstructorValidation::test_min_node_split_size_validation
[gw2] [ 11%] PASSED tests/test_mice.py::TestConstructorValidation::test_min_node_split_size_validation
tests/test_mice.py::TestConstructorValidation::test_adjustment_parameters_require_monotone
tests/test_mice.py::TestConstructorValidation::test_monotone_validation
tests/test_mice.py::TestConstructorValidation::test_x_idx_validation
[gw2] [ 13%] PASSED tests/test_mice.py::TestConstructorValidation::test_adjustment_parameters_require_monotone
[gw3] [ 14%] PASSED tests/test_mice.py::TestConstructorValidation::test_monotone_validation
[gw1] [ 16%] PASSED tests/test_mice.py::TestConstructorValidation::test_x_idx_validation
tests/test_mice.py::TestBasicImputation::test_imputation_preserves_non_missing_values
tests/test_mice.py::TestBasicImputation::test_impute_validates_input_shape
tests/test_mice.py::TestConstructorValidation::test_adjustment_parameter_lengths
[gw2] [ 18%] PASSED tests/test_mice.py::TestConstructorValidation::test_adjustment_parameter_lengths
tests/test_mice.py::TestConstructorValidation::test_properties
[gw2] [ 19%] PASSED tests/test_mice.py::TestConstructorValidation::test_properties
tests/test_mice.py::TestBasicImputation::test_simple_imputation_single_missing_value
[gw1] [ 21%] PASSED tests/test_mice.py::TestBasicImputation::test_impute_validates_input_shape
tests/test_mice.py::TestBasicImputation::test_impute_validates_missing_pattern
[gw1] [ 22%] PASSED tests/test_mice.py::TestBasicImputation::test_impute_validates_missing_pattern
tests/test_mice.py::TestMonotoneMode::test_monotone_mode_basic
[gw2] [ 24%] PASSED tests/test_mice.py::TestBasicImputation::test_simple_imputation_single_missing_value
[gw1] [ 26%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_mode_basic
tests/test_mice.py::TestMonotoneMode::test_non_monotone_mode_performs_initialization
tests/test_mice.py::TestMonotoneMode::test_monotone_detects_non_monotone_pattern
[gw1] [ 27%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_detects_non_monotone_pattern
tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_scale_adjustment
[gw2] [ 29%] PASSED tests/test_mice.py::TestMonotoneMode::test_non_monotone_mode_performs_initialization
tests/test_mice.py::TestMonotoneMode::test_monotone_mode_skips_initialization
[gw2] [ 31%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_mode_skips_initialization
[gw1] [ 32%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_scale_adjustment
tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_more_zeros_adjustment
tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_shift_adjustment
[gw2] [ 34%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_shift_adjustment
[gw1] [ 36%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_more_zeros_adjustment
tests/test_mice.py::TestReproducibility::test_same_seed_produces_identical_results
tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_fewer_zeros_adjustment
[gw3] [ 37%] PASSED tests/test_mice.py::TestBasicImputation::test_imputation_preserves_non_missing_values
tests/test_mice.py::TestBasicImputation::test_all_missing_values_imputed
[gw1] [ 39%] PASSED tests/test_mice.py::TestMonotoneMode::test_monotone_mode_with_fewer_zeros_adjustment
tests/test_mice.py::TestReproducibility::test_statistics_reproducible_with_same_seed
[gw3] [ 40%] PASSED tests/test_mice.py::TestBasicImputation::test_all_missing_values_imputed
tests/test_mice.py::TestBasicImputation::test_imputation_with_ignored_variables
[gw2] [ 42%] PASSED tests/test_mice.py::TestReproducibility::test_same_seed_produces_identical_results
tests/test_mice.py::TestReproducibility::test_different_seeds_produce_different_results
[gw1] [ 44%] PASSED tests/test_mice.py::TestReproducibility::test_statistics_reproducible_with_same_seed
tests/test_mice.py::TestReproducibility::test_seed_increment_across_iterations
[gw3] [ 45%] PASSED tests/test_mice.py::TestBasicImputation::test_imputation_with_ignored_variables
tests/test_mice.py::TestBasicImputation::test_get_ival_stats_shape
[gw1] [ 47%] PASSED tests/test_mice.py::TestReproducibility::test_seed_increment_across_iterations
tests/test_misc.py::test_population
[gw2] [ 49%] PASSED tests/test_mice.py::TestReproducibility::test_different_seeds_produce_different_results
tests/test_mice.py::TestReproducibility::test_multiple_imputation_with_different_seeds
[gw3] [ 50%] PASSED tests/test_mice.py::TestBasicImputation::test_get_ival_stats_shape
tests/test_reweight.py::test_drop_impossible_targets_removes_all_zero_column
[gw3] [ 52%] PASSED tests/test_reweight.py::test_drop_impossible_targets_removes_all_zero_column
tests/test_reweight.py::test_drop_impossible_targets_keeps_all_when_none_zero
[gw3] [ 54%] PASSED tests/test_reweight.py::test_drop_impossible_targets_keeps_all_when_none_zero
tests/test_reweight.py::test_drop_impossible_targets_column_with_single_nonzero_is_kept
[gw3] [ 55%] PASSED tests/test_reweight.py::test_drop_impossible_targets_column_with_single_nonzero_is_kept
tests/test_reweight.py::test_no_all_zero_columns_in_real_loss_matrix
[gw2] [ 57%] PASSED tests/test_mice.py::TestReproducibility::test_multiple_imputation_with_different_seeds
tests/test_misc.py::test_no_negative_weights
[gw1] [ 59%] PASSED tests/test_misc.py::test_population
tests/test_revenue_levels_cbo.py::test_revenue_levels_cbo
[gw2] [ 60%] PASSED tests/test_misc.py::test_no_negative_weights
tests/test_reweight.py::test_negative_target_tolerance_on_abs
[gw2] [ 62%] PASSED tests/test_reweight.py::test_negative_target_tolerance_on_abs
tests/test_reweight.py::test_tolerance_override_widens_constraint
[gw2] [ 63%] PASSED tests/test_reweight.py::test_tolerance_override_widens_constraint
tests/test_soi_sanity_2022.py::test_soi_sanity_2022
[gw2] [ 65%] PASSED tests/test_soi_sanity_2022.py::test_soi_sanity_2022
tests/test_solve_weights.py::test_solver_xx
[gw3] [ 67%] PASSED tests/test_reweight.py::test_no_all_zero_columns_in_real_loss_matrix
tests/test_reweight.py::test_positive_target_symmetric_band
[gw3] [ 68%] PASSED tests/test_reweight.py::test_positive_target_symmetric_band
tests/test_solve_weights.py::test_parse_log_solved
[gw3] [ 70%] PASSED tests/test_solve_weights.py::test_parse_log_solved
tests/test_solve_weights.py::test_parse_log_missing
[gw3] [ 72%] PASSED tests/test_solve_weights.py::test_parse_log_missing
tests/test_solve_weights.py::test_solve_weights_parse_scope_states
[gw3] [ 73%] PASSED tests/test_solve_weights.py::test_solve_weights_parse_scope_states
tests/test_solve_weights.py::test_solve_weights_parse_scope_specific
[gw3] [ 75%] PASSED tests/test_solve_weights.py::test_solve_weights_parse_scope_specific
tests/test_solve_weights.py::test_solve_weights_parse_scope_excludes
[gw3] [ 77%] PASSED tests/test_solve_weights.py::test_solve_weights_parse_scope_excludes
tests/test_solve_weights.py::test_filter_areas_states
[gw3] [ 78%] PASSED tests/test_solve_weights.py::test_filter_areas_states
tests/test_solve_weights.py::test_filter_areas_cds
[gw3] [ 80%] PASSED tests/test_solve_weights.py::test_filter_areas_cds
tests/test_solve_weights.py::test_filter_areas_specific
[gw3] [ 81%] PASSED tests/test_solve_weights.py::test_filter_areas_specific
tests/test_solve_weights.py::test_humanize_desc
[gw3] [ 83%] PASSED tests/test_solve_weights.py::test_humanize_desc
tests/test_tax_expenditures.py::test_tax_exp_diffs
[gw3] [ 85%] FAILED tests/test_tax_expenditures.py::test_tax_exp_diffs
tests/test_tmd_file_fingerprint.py::test_tmd_file_fingerprint
[gw1] [ 86%] PASSED tests/test_revenue_levels_cbo.py::test_revenue_levels_cbo
tests/test_reweight.py::test_zero_target_exact_bounds
[gw1] [ 88%] PASSED tests/test_reweight.py::test_zero_target_exact_bounds
[gw3] [ 90%] FAILED tests/test_tmd_file_fingerprint.py::test_tmd_file_fingerprint
tests/test_weights.py::test_weights
[gw3] [ 91%] FAILED tests/test_weights.py::test_weights
[gw0] [ 93%] FAILED tests/test_imputed_variables.py::test_obbba_deduction_tax_benefits
tests/test_imputed_variables.py::test_imputed_variable_distribution
[gw0] [ 95%] SKIPPED tests/test_imputed_variables.py::test_imputed_variable_distribution
tests/test_mice.py::TestConstructorValidation::test_valid_minimal_constructor
[gw0] [ 96%] PASSED tests/test_mice.py::TestConstructorValidation::test_valid_minimal_constructor
[gw2] [ 98%] PASSED tests/test_solve_weights.py::test_solver_xx
tests/test_solve_weights.py::test_solver_xx_targets_hit
[gw2] [100%] PASSED tests/test_solve_weights.py::test_solver_xx_targets_hit

================================================================================ FAILURES =================================================================================
___________________________________________________________________________ test_tax_exp_diffs ____________________________________________________________________________

start_year = 2022, gfactors = WindowsPath('tmd/storage/output/tmd_growfactors.csv')
weights = WindowsPath('tmd/storage/output/tmd_weights.csv.gz'), weights_scale = 1.0

            if self.__aging_data:
                if not isinstance(gfactors, GrowFactors):
>                   raise ValueError('gfactors is not a GrowFactors instance')
E                   ValueError: gfactors is not a GrowFactors instance

taxcalc\data.py:103: ValueError
________________________________________________________________________ test_tmd_file_fingerprint ________________________________________________________________________

E           Failed: 34 statistic mismatch(es) in tmd.csv.gz fingerprint:
E           auto_loan_interest.sum: 1.1897e+08 -> 1.20033e+08 (rel diff 8.94e-03, rtol=1e-03)
E           auto_loan_interest.weighted_sum: 9.33899e+10 -> 9.50462e+10 (rel diff 1.77e-02, rtol=1e-03)
E           auto_loan_interest.std: 1588.18 -> 1602.84 (rel diff 9.23e-03, rtol=1e-03)
E           e00200s.std: 1.40759e+06 -> 1.39983e+06 (rel diff 5.52e-03, rtol=1e-03)
E           e00900p.sum: 4.48133e+09 -> 4.49111e+09 (rel diff 2.18e-03, rtol=1e-03)
E           e00900p.weighted_sum: 3.1539e+11 -> 3.15882e+11 (rel diff 1.56e-03, rtol=1e-03)
E           e00900p.std: 506638 -> 504426 (rel diff 4.37e-03, rtol=1e-03)
E           e00900s.sum: 1.37155e+09 -> 1.36178e+09 (rel diff 7.13e-03, rtol=1e-03)
E           e00900s.weighted_sum: 9.60751e+10 -> 9.55834e+10 (rel diff 5.12e-03, rtol=1e-03)
E           e00900s.std: 198753 -> 203938 (rel diff 2.61e-02, rtol=1e-03)
E           e02100p.weighted_sum: 5.70894e+09 -> 5.71583e+09 (rel diff 1.21e-03, rtol=1e-03)
E           e02100s.sum: 1.73129e+08 -> 1.72947e+08 (rel diff 1.05e-03, rtol=1e-03)
E           e02100s.weighted_sum: 1.67602e+09 -> 1.66929e+09 (rel diff 4.01e-03, rtol=1e-03)
E           n1820.sum: 8747 -> 8692 (rel diff 6.29e-03, rtol=1e-03)
E           n1820.weighted_sum: 5.86428e+06 -> 6.00202e+06 (rel diff 2.35e-02, rtol=1e-03)
E           n1820.std: 0.204522 -> 0.203058 (rel diff 7.16e-03, rtol=1e-03)
E           n21.sum: 13308 -> 13284 (rel diff 1.80e-03, rtol=1e-03)
E           n21.weighted_sum: 1.15454e+07 -> 1.15045e+07 (rel diff 3.55e-03, rtol=1e-03)
E           nu06.weighted_sum: 3.26012e+07 -> 3.26748e+07 (rel diff 2.26e-03, rtol=1e-03)
E           nu18.weighted_sum: 7.9834e+07 -> 7.97384e+07 (rel diff 1.20e-03, rtol=1e-03)
E           overtime_income.sum: 2.45115e+09 -> 2.37805e+09 (rel diff 2.98e-02, rtol=1e-03)
E           overtime_income.weighted_sum: 2.53143e+11 -> 2.59648e+11 (rel diff 2.57e-02, rtol=1e-03)
E           overtime_income.std: 286624 -> 213247 (rel diff 2.56e-01, rtol=1e-03)
E           overtime_income.max: 6.80801e+07 -> 2.5051e+07 (rel diff 6.32e-01, rtol=1e-03)
E           pencon_p.sum: 6.90867e+08 -> 5.99531e+08 (rel diff 1.32e-01, rtol=1e-03)
E           pencon_p.weighted_sum: 2.63782e+11 -> 2.73884e+11 (rel diff 3.83e-02, rtol=1e-03)
E           pencon_p.std: 6385.95 -> 5491.71 (rel diff 1.40e-01, rtol=1e-03)
E           pencon_s.sum: 3.76014e+08 -> 3.45125e+08 (rel diff 8.21e-02, rtol=1e-03)
E           pencon_s.weighted_sum: 7.66177e+10 -> 7.96429e+10 (rel diff 3.95e-02, rtol=1e-03)
E           pencon_s.std: 4874.63 -> 4374.53 (rel diff 1.03e-01, rtol=1e-03)
E           tip_income.sum: 4.54325e+08 -> 4.24166e+08 (rel diff 6.64e-02, rtol=1e-03)
E           tip_income.weighted_sum: 6.5982e+10 -> 6.49453e+10 (rel diff 1.57e-02, rtol=1e-03)
E           tip_income.std: 103060 -> 126101 (rel diff 2.24e-01, rtol=1e-03)
E           tip_income.max: 2.41441e+07 -> 4.86025e+07 (rel diff 1.01e+00, rtol=1e-03)
E
E           Tax-Calculator version drift: fingerprint generated under 6.5.2, currently installed 6.6.2. Some TMD outputs are downstream of Tax-Calculator behavior, so a version change can legitimately shift these statistics. If the new version is intentional, regenerate the reference with: pytest tests/test_tmd_file_fingerprint.py --update-fingerprint

tests\test_tmd_file_fingerprint.py:234: Failed
______________________________________________________________________________ test_weights _______________________________________________________________________________

E           ValueError: WEIGHT FINGERPRINT MISMATCH USING 2022 DATA:
E             p50: actual=398.079155, expected=398.097

tests\test_weights.py:83: ValueError
____________________________________________________________________ test_obbba_deduction_tax_benefits ____________________________________________________________________

E           ValueError:
E           IMPUTED VARIABLE DEDUCTION BENEFIT ACT-vs-EXP DIFFS USING 2022 DATA:
E           DIFF:OTM,totben,act,exp,atol= 24.11 23.71 0.01
E           DIFF:OTM,affpct,act,exp,atol= 8.58 8.56 0.01
E           DIFF:OTM,affben,act,exp,atol= 1435.0 1413 1.0
E           DIFF:TIP,totben,act,exp,atol= 6.93 7.11 0.01
E           DIFF:TIP,affpct,act,exp,atol= 2.53 2.59 0.01
E           DIFF:TIP,affben,act,exp,atol= 1398.0 1401 1.0
E           DIFF:ALI,totben,act,exp,atol= 8.06 7.87 0.01
E           DIFF:ALI,affpct,act,exp,atol= 12.25 12.28 0.01
E           DIFF:ALI,affben,act,exp,atol= 336.0 327 1.0
E           DIFF:ALL,totben,act,exp,atol= 59.45 58.97 0.01
E           DIFF:ALL,affpct,act,exp,atol= 28.17 28.24 0.01
E           DIFF:ALL,affben,act,exp,atol= 1077.0 1065 1.0

tests\test_imputed_variables.py:175: ValueError
========================================================================= short test summary info =========================================================================
FAILED tests/test_tax_expenditures.py::test_tax_exp_diffs - ValueError: gfactors is not a GrowFactors instance
FAILED tests/test_tmd_file_fingerprint.py::test_tmd_file_fingerprint - Failed: 34 statistic mismatch(es) in tmd.csv.gz fingerprint:
FAILED tests/test_weights.py::test_weights - ValueError: WEIGHT FINGERPRINT MISMATCH USING 2022 DATA:
FAILED tests/test_imputed_variables.py::test_obbba_deduction_tax_benefits - ValueError:
=========================================================== 4 failed, 56 passed, 1 skipped in 76.83s (0:01:16) ============================================================
make: *** [test] Error 1

--- a/tmd/utils/taxcalc_output.py
+++ b/tmd/utils/taxcalc_output.py
@@ -29,11 +29,11 @@ def add_taxcalc_outputs(
     Returns:
         pd.DataFrame: The Tax-Calculator output.
     """
-    if isinstance(weights, pathlib.PosixPath):
+    if isinstance(weights, pathlib.Path):
         wghts = str(weights)
     else:
         wghts = weights
-    if isinstance(growfactors, pathlib.PosixPath):
+    if isinstance(growfactors, pathlib.Path):
         growf = taxcalc.GrowFactors(growfactors_filename=str(growfactors))
     else:
         growf = growfactors

